gunicorn/config.py

```diff
@@ -344,6 +344,23 @@ def validate_pos_int(val):
     return val
 
 
+def validate_ssl_version(val):
+    ssl_versions = {}
+    for protocol in [p for p in dir(ssl) if p.startswith("PROTOCOL_")]:
+        ssl_versions[protocol[9:]] = getattr(ssl, protocol)
+    if val in ssl_versions:
+        return ssl_versions[val]
+    else:
+        try:
+            intval = validate_pos_int(val)
+            if intval in ssl_versions.values():
+                return intval
+        except:
```

Comment:
I don't think we should silence exceptions here. `validate_pos_int()` can raise both `TypeError` and `ValueError` and IMO both are expected to be raised. Also, it would be nice to have some tests to cover this part of the code.

Reply:
Looking for some advice on how to change this.

If none of the following are passed:
- String matching trailing part of `PROTOCOL_` key
- Positive integer which also matches a value from the `PROTOCOL_` map

... the PR code as it stands will raise a `ValueError` with the custom valid-options iteration.  If a string is passed which doesn't match a `PROTOCOL_`, it will fail integer conversion with a `TypeError`, which as you point out is currently dropped in favour of the custom `ValueError`.  Negative numbers produce a `ValueError` which is likewise dropped in favour of the custom valid-options `ValueError`.  Positive numbers that fail to match just fall-through to the custom exception.

Options I can see for not dropping the exception from `validate_pos_int`:

1. Only catch `ValueError` and allow `TypeError` to propagate, but that means `--ssl_version foo` won't generate the valid options message.
1. Catch both `ValueError` and `TypeError` and instead of `pass`, raise the custom exception there.
1. Use Python 3 exception-chaining and wrap the thrown `ValueError` and `TypeError` in the custom exception.

Reply:
Thank you for the great analysis. I think option 2 is the best choice.